core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -325,24 +309,50 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
       case Meta(c, data)    => c.stripFailures.map(Meta(_, data))
     }
 
-  final def succeeded: Boolean = !failed
-
   final def traces: List[ZTrace] =
     self
       .foldLeft(List.empty[ZTrace]) {
         case (z, Traced(_, trace)) => trace :: z
       }
       .reverse
 
-  private def foldLeft[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z =
-    (f.lift(z -> self).getOrElse(z), self) match {
-      case (z, Then(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Both(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Traced(cause, _))  => cause.foldLeft(z)(f)
-      case (z, Meta(cause, _))    => cause.foldLeft(z)(f)
+  private def find[Z](f: PartialFunction[Cause[E], Z]): Option[Z] = {
+    @tailrec
+    def loop(cause: Cause[E], stack: List[Cause[E]]): Option[Z] =
+      f.lift(cause) match {
+        case Some(z) => Some(z)
+        case None =>
+          cause match {
+            case Then(left, right) => loop(left, right :: stack)
+            case Both(left, right) => loop(left, right :: stack)
+            case Traced(cause, _)  => loop(cause, stack)
+            case Meta(cause, _)    => loop(cause, stack)
+            case _ =>
+              stack match {
+                case hd :: tl => loop(hd, tl)
+                case Nil      => None
+              }
+          }
+      }
+    loop(self, Nil)
+  }
 
-      case (z, _) => z
-    }
+  private def foldLeft[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z = {
+    @tailrec
+    def loop(z: Z, cause: Cause[E], stack: List[Cause[E]]): Z =
+      (f.lift(z -> cause).getOrElse(z), cause) match {
```

Comment:
You may want to memoize `f.lift` as it create a new object on each invocation.

Reply:
In fact this is just `f.applyOrElse(z -> cause, _ => z)` in hiding, without the `Option` allocation

Reply:
Good point. Will update.